File: MoonReader/Models/ReadingStatistics.swift
//
//  ReadingStatistics.swift
//  MoonReader
//
//  Reading statistics model - tương đương statistics table trong Android
//

import Foundation

struct ReadingStatistics: Identifiable, Codable {
    var id: UUID
    var bookFilename: String
    var usedTime: TimeInterval // seconds
    var readWords: Int
    var dates: [Date]
    
    init(
        id: UUID = UUID(),
        bookFilename: String,
        usedTime: TimeInterval = 0,
        readWords: Int = 0,
        dates: [Date] = []
    ) {
        self.id = id
        self.bookFilename = bookFilename
        self.usedTime = usedTime
        self.readWords = readWords
        self.dates = dates
    }
    
    var readHours: Double {
        return usedTime / 3600.0
    }
    
    var averageWordsPerMinute: Int {
        guard usedTime > 0 else { return 0 }
        return Int(Double(readWords) / (usedTime / 60.0))
    }
}

File: MoonReaderRN/src/models/ReadingStatistics.ts
import { Model } from '@nozbe/watermelondb';
import { field, text } from '@nozbe/watermelondb/decorators';

// TypeScript interface
export interface ReadingStatistics {
  id: string;
  bookFilename: string;
  usedTime: number; // seconds
  readWords: number;
  dates: Date[];
}

// WatermelonDB Model class
export class ReadingStatisticsModel extends Model {
  static table = 'statistics';

  @text('bookFilename') bookFilename!: string;
  @field('usedTime') usedTime!: number;
  @field('readWords') readWords!: number;
  @text('dates') dates!: string; // JSON array as string

  // Helper methods
  get readHours(): number {
    return this.usedTime / 3600.0;
  }

  get averageWordsPerMinute(): number {
    if (this.usedTime === 0) return 0;
    return Math.round(this.readWords / (this.usedTime / 60.0));
  }

  get datesArray(): Date[] {
    try {
      const timestamps = JSON.parse(this.dates || '[]') as number[];
      return timestamps.map((t) => new Date(t));
    } catch {
      return [];
    }
  }
}

